pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -0,0 +1,339 @@
+
+		// For now, this will be identical to the not found path.  When the SF iterator datastore
+		// is fully implemented, the found path will look up the shared iterator and create
+		// a clone from it.
+	} else {
+		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
+			sf.internalStorage.mu.Unlock()
+			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationRead).Inc()
+
+			// we cannot share this iterator because we have reached the size limit.
+			return sf.RelationshipTupleReader.Read(ctx, store, tupleKey, options)
+		}
+
+		sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
+			counter: 1,
+		}
+		sf.internalStorage.mu.Unlock()
+		span.SetAttributes(attribute.Bool("found", false))
+	}
+	actual, err := sf.RelationshipTupleReader.Read(ctx, store, tupleKey, options)
+	if err != nil {
+		go func() {
+			_, _ = sf.deref(cacheKey)
+		}()
+		return nil, err
+	}
+	sharedIteratorQueryHistogram.WithLabelValues(
+		storagewrappersutil.OperationRead, strconv.FormatBool(found),
+	).Observe(float64(time.Since(start).Milliseconds()))
+
+	return newSharedIterator(sf, cacheKey, actual), nil
+}
+
+// decrement cacheKey from internal reference count
+// If reference count is 0, remove cacheKey.
+// Return whether item is removed.
+func (sf *IteratorDatastore) deref(cacheKey string) (bool, error) {
+	sf.internalStorage.mu.Lock()
+	defer sf.internalStorage.mu.Unlock()
+	item, ok := sf.internalStorage.iters[cacheKey]
+	if !ok {
+		sf.logger.Error("failed to dereference cache key", zap.String("cacheKey", cacheKey))
+		return false, fmt.Errorf("failed to dereference cache key: %s", cacheKey)
+	}
+	item.counter--
+	if item.counter == 0 {
+		delete(sf.internalStorage.iters, cacheKey)
+		return true, nil
+	}
+	return false, nil
+}
+
+// sharedIterator will be shared with multiple consumers.
+// For now, this is a skeleton and no one is actually sharing the iterator.
+type sharedIterator struct {
+	manager *IteratorDatastore
+	key     string
+	inner   storage.TupleIterator
+	stopped bool
+}
+
+func newSharedIterator(manager *IteratorDatastore, key string, inner storage.TupleIterator) *sharedIterator {
+	return &sharedIterator{
+		manager: manager,
+		key:     key,
+		inner:   inner,
+		stopped: false,
+	}
+}
+
+func (s *sharedIterator) Head(ctx context.Context) (*openfgav1.Tuple, error) {
+	return s.inner.Head(ctx)
+}
+
+func (s *sharedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
+	return s.inner.Next(ctx)
+}
+
+func (s *sharedIterator) Stop() {
+	// For now, we don't check whether this is the last item in the shared iterator
+	// as this is just a skeleton.  When the actual shared iterator is implemented,
+	// we will stop the inner iterator only if this is the last reference.
+	if !s.stopped {
+		// It is perfectly possible that iterator calling stop more than once.
+		// However, we only want to decrement the count on the first stop.
+		_, _ = s.manager.deref(s.key)
+	}
+	s.stopped = true
```

Comment:
Consider making `s.stopped` an `atomic.Bool`:
```suggestion
	if !s.stopped.Swap(true) {
		// It is perfectly possible that iterator calling stop more than once.
		// However, we only want to decrement the count on the first stop.
		_, _ = s.manager.deref(s.key)
	}
```